tls: throw an error on getLegacyCipher

jasnell opened the pull request:
throw an error on unknown getLegacyCipher input
rather than returning the default. Makes the
behavior more explicit and matches up with the
behavior on the command line switches.

jasnell commented:
@misterdjules @mdawsonibm

misterdjules commented:
Can we add a test to make sure that `getLecagyCiphers` throws/doesn't throw as expected?

jasnell commented:
Done

jasnell commented:
https://github.com/joyent/node/issues/15445 @misterdjules

mdawsonibm commented:
One minor nit.  Should the tests validate the type of the Error, ie a TypeError or an Error as appropriate ? Otherwise lgtm.

mdawsonibm commented:
One minor nit.  Should the tests validate the type of the Error, ie a TypeError or an Error as appropriate ? Otherwise lgtm.

jasnell commented:
/cc @mhdawson @misterdjules ok, please check the two additional commits. if these look good I'll squash the pr down and land.

jasnell committed:
tls: order of precedence on cipher list options, type checking

This commit squashes down several previous commits on this change
to clean up before landing. There are several changes to the
new cipher list command line switches:

1. It includes the type checking fix on getLegacyCiphers
2. It has the command line switches override the environment
   variables
3. It documents it order of precedence
4. Expands the test case to test for order of precedence

jasnell committed:
test: test-tls-cipher-list test error type

test that the type of error thrown is correct

jasnell committed:
tls: do not set default ciphers if v0.10.38 cipher list is set

for consistency with the v0.10.38, if the v0.10.38 cipher list
is being used, do not set the default ciphers on the options
in connect.

jasnell commented:
@mhdawson @misterdjules ok, made the additional edits discussed and squashed the commits. Please verify.

jasnell committed:
test: fix typo in tls.markdown doc and additional tests

Fix a minor typo in the tls.markdown and add additional
precedence tests in test-tls-cipher-list

misterdjules commented on lib/_tls_wrap.js:
```diff
@@ -836,10 +836,13 @@ exports.connect = function(/* [port, host], options, cb */) {
     checkServerIdentity: tls.checkServerIdentity
   };
 
+  if (tls.DEFAULT_CIPHERS != tls.getLegacyCiphers('v0.10.38')) {
```
If one passes `--cipher-list='ECDHE-RSA-AES128-SHA256:AES128-GCM-SHA256:RC4:HIGH:!MD5:!aNULL:!EDH'`, my understanding is that `tls.DEFAULT_CIPHERS` will === `tls.getLegacyCiphers('v0.10.38'))`, and `default.ciphers` won't be set. I haven't tested it, so it may be a wrong assumption.

So I guess that here what is meant is more `if (user didn't pass --enable-legacy-cipher-list=v0.10.38) { ... }`, which is not equivalent to `if (tls.DEFAULT_CIPHERS != tls.getLegacyCiphers('v0.10.38'))`.

Thoughts?

misterdjules replied:
Also, in general please use strict equality unless not possible. In this occurence, please use `!==` instead of `!=`.

jasnell replied:
Yes, that is the case. I personally don't see this as a significant problem. If someone passes in the exact default cipher list as v0.10.38 then the likely intent is to revert the behavior back to v0.10.38, which means the default.ciphers not being set. It's an acceptable edge case IMHO but I'm good either way. It's a bit of a larger change tho since this does not yet specifically track whether `--enable-legacy-cipher-list` is passed vs. `--cipher-list`

misterdjules commented on test/simple/test-tls-cipher-list.js:
```diff
@@ -40,6 +40,54 @@ function doTest(checklist, env, useswitch) {
       // Test NODE_LEGACY_CIPHER_LIST
       if (env) options = {env:{"NODE_LEGACY_CIPHER_LIST": env}};
       break;
+    case 4:
```
Having arbitrary numbers determine the behavior of tests seems very confusing to me, because it's not clear when I read `doTest(V1038Ciphers, 'v0.10.38', 8);` that we test that `NODE_LEGACY_CIPHER_LIST takes precendence over NODE_CIPHER_LIST`.

Can we refactor that to make it clearer? Maybe we could have separate function clearly named for each test we want to perform? Parameters for each test would be local to these functions and make it clear what each test does.

misterdjules commented on doc/api/tls.markdown:
```diff
@@ -223,10 +252,15 @@ Example:
+release version. The only values currently supported are `v0.10.38`, `v0.10.39`, `v0.12.2` and `v0.12.3`.
 
-Example:
-    var cipher_suite = tls.getLegacyCiphers('v0.10.38');
```
Keeping an example (with code formatting) would be nice.

jasnell committed:
tls: refactor test-tls-cipher-list and v0.10.38 legacy test

Per the feedback from Julien, refactor the test-tls-cipher-list
test to avoid using the confusing switch statement. Add tests
to ensure that using `--enable-legacy-cipher-list=v0.10.38`
or `NODE_LEGACY_CIPHER_LIST=v0.10.38` disables the use of the
default cipher list but setting the `--cipher-list` equal to
the v0.10.38 list explicitly does not.

Minor refactor to tls.js to ensure the above behavior.

jasnell committed:
test: additional refinements in the test

Per the latest round of feedback from julien.
Fix a couple of typos, add some explanatory text,
refactor the test case.

jasnell committed:
test: additional round of edits on the test-tls-cipher-list.js

Per feedback from Julien.

jasnell committed:
test: refactor default cipher list test

Refactor the default cipher list test to separate
concerns a bit, make the test clearer based on
Julien's feedback.

jasnell commented:
Closing this as it cannot land here now. Will be refactoring this change and opening a new PR against master on nodejs/node.
